aiida/manage/configuration/config.py

```diff
@@ -18,12 +18,34 @@
 
 from aiida.common import json
 
-from .migrations import CURRENT_CONFIG_VERSION, OLDEST_COMPATIBLE_CONFIG_VERSION
+from .migrations import CURRENT_CONFIG_VERSION, OLDEST_COMPATIBLE_CONFIG_VERSION, get_current_version
 from .options import get_option, parse_option, NO_DEFAULT
 from .profile import Profile
 from .settings import DEFAULT_UMASK, DEFAULT_CONFIG_INDENT_SIZE
 
-__all__ = ('Config',)
+__all__ = ('Config', 'create_backup')
+
+
+def create_backup(config_dictionary, filepath):
+    """Create backup of configuration.
+
+    :param config_dictionary: contents of the configuration file
+    :param filepath: absolute path to the current configuration file
+    """
+    from aiida.common import timezone
+    from aiida.cmdline.utils import echo
+
+    version = get_current_version(config_dictionary)
+    filepath_backup = '{}.v{}.{}'.format(filepath, version, timezone.now().strftime('%Y%m%d-%H%M%S'))
```

Comment:
Maybe switch the order of version and date so that the files will be ordered by date.

Also, we already have a backup function that creates backups of the config file ending in `.json~`
https://github.com/aiidateam/aiida-core/blob/cd916288509f31ad088120f24193bb2c42c6afc2/aiida/manage/configuration/config.py#L336

Perhaps we should join with this function.

Reply:
Can't do because that is a member function of `Config` which in order to be constructed requires a fully migrated config dictionary, so at that point creating a backup will be too late. It would be possible if we move the logic that determines whether migration is necessary in the constructor of the `Config` class. I am not sure if that is better. The code there is already a bit complicated and might make it even more tightly coupled. On the other hand maybe it makes it more self-contained and therefore better. I will take a look what that would look like.

Reply:
Also re the ordering, with the current order they will still be ordered by data, within each version group, which should also be ordered, no?

Reply:
Decided to omit the version as it is in the files themselves.